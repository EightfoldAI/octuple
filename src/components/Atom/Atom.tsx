'use client';

import React from 'react';
import { mergeClasses } from '../../shared/utilities';
import { AtomProps } from './Atom.types';

export const Atom = React.forwardRef(
  ({ of, children, classes = [], ...props }, ref): JSX.Element => {
    const computedClassName = mergeClasses([...classes, 'oct-component']);
    return React.createElement(
      of,
      { ...props, ref, className: computedClassName },
      children
    );
  }
) as <T, U = any>(props: AtomProps<T, U>) => JSX.Element;
