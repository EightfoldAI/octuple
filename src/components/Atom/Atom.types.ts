import React from 'react';
import { ArgumentArray } from '../../shared/utilities';

type OcAtomBaseProps<T> = {
  /**
   * Custom classnames of the component
   */
  classNames?: string;
  /**
   * Unique id used to target element for testing
   */
  'data-test-id'?: string;

  style?: React.CSSProperties;

  ref?: React.Ref<T>;
};

export type AtomProps<T = {}, U = React.Ref<any>> = Omit<T, 'children'> &
  OcAtomBaseProps<U> & {
    of: keyof HTMLElementTagNameMap;
    classes: ArgumentArray;
    children?: React.ReactNode;
  };
