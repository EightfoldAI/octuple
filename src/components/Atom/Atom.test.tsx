import React, { useMemo, useRef } from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { Atom } from '../Atom';
import { render } from '@testing-library/react';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

const AtomComponent = (props: {
  classNames?: string;
  style?: React.CSSProperties;
}): JSX.Element => {
  const atomRef: React.MutableRefObject<HTMLDivElement> =
    useRef<HTMLDivElement>(null);

  useMemo(() => {
    if (atomRef.current) {
      atomRef.current.id = 'ref-test';
    }
  }, []);

  return (
    <Atom<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>
      of="div"
      ref={atomRef}
      classes={[props.classNames]}
      style={props.style}
      data-test-id="test-atom-legacy"
      data-testid="test-atom"
    >
      Test Atom component.
    </Atom>
  );
};

describe('Atom', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Renders without crashing and uses data-testid', () => {
    const { container, getByTestId } = render(<AtomComponent />);
    const atom = getByTestId('test-atom');
    expect(() => container).not.toThrowError();
    expect(atom).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('Atom accepts custom classes', () => {
    const { container } = render(<AtomComponent classNames="atom-class" />);
    expect(container.querySelector('.atom-class')).toBeTruthy();
  });

  test('Atom accepts custom styles', () => {
    const { container, getByTestId } = render(
      <AtomComponent style={{ backgroundColor: 'red' }} />
    );
    const atom = getByTestId('test-atom');
    expect(atom.style.backgroundColor).toBe('red');
  });

  test('Atom accepts legacy data-test-id', () => {
    const { container } = render(<AtomComponent classNames="atom-class" />);
    expect(
      container
        .getElementsByClassName('atom-class')[0]
        .getAttribute('data-test-id')
    ).toBe('test-atom-legacy');
  });
});
