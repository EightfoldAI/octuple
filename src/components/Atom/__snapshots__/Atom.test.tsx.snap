// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom Renders without crashing and uses data-testid 1`] = `
<div>
  <div
    class="oct-component"
    data-test-id="test-atom-legacy"
    data-testid="test-atom"
  >
    Test Atom component.
  </div>
</div>
`;
